teps/0097-breakpoints-for-taskruns-and-pipelineruns.md

```diff
@@ -0,0 +1,317 @@
+    - [Debug Environment Additions ](#debug-environment-additions)
+        - [Mounts](#mounts)
+        - [Debug Scripts](#debug-scripts)
+- [PipelineRun Proposal](#pipelinerun-proposal)
+    - [Breakpoint before/after a TaskRun](#breakpoint-before-or-after-a-taskrun)
+        - [before TaskRun](#before-taskrun)
+        - [after TaskRun](#after-taskrun)
+    - [Breakpoint on failure of a TaskRun](#breakpoint-on-failure-of-a-taskrun)
+- [User Stories](#user-stories)
+    - [CLI integration](#cli-integration)
+        - [Environment access](#environment-access)
+- [Alternatives](#alternatives)
+-<!-- /toc -->
+## Summary
+
+Debugging TaskRuns can be tiresome. Re-running Tasks to figure out which part of a particular step is to blame would tax productivity.
+By enabling breakpoint on failure for a TaskRun we should be able to halt a TaskRun at the failing step and get access to the
+step environment to analyze cause of the failure. Unlike legacy systems, this allows the user to debug TaskRuns while the TaskRun 
+is still running, hence improving developer productivity and Pipeline debuggability.
+
+## Motivation
+
+Lack of live debugging for Pipelines.
+
+### Goals
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata: 
+  generateName: quarante-deux-deux
+spec:
+  debug:
+    breakpoint:
+      onFailure: true
+      beforeStep: ["task-build-archive"]
+      afterStep: ["deploy-frontend"]
+```
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: TaskRun
+metadata:
+  generateName: quarante-deux
+spec:
+  debug:
+    breakpoint:
+      onFailure: true
+      beforeStep: ["test-js"]
+      afterStep: ["push-to-registry"]
+```
+
+### Use Cases (optional)
+
+This enhancement can allow a more interactive way of working with TaskRuns. CLI or IDE extensions could allow users to run a 
+Task with breakpoint-on-failure and while the logs are being streamed, the user can be dropped to a shell in the breakpoint
+failed step and allow the user to debug as they develop.
+
+## Requirements
+
+* To enable breakpoint on failure of a step, it would be necessary to update the Entrypoint to support lifecycle
+changes when breakpoint on failure has been enabled.
+
+* To enable breakpoint after a step, similar changes need to be made as the step above.
+
+* To enable breakpoint before a step, we need to let the step know before execution that there is a breakpoint set.
+
+* The step environment would have to have scripts which the user can use to clear the breakpoint or
+mark the step as a success or a failure.
+
+## TaskRun Proposal
+
+### Breakpoint before/after a step
+The `breakpoint` spec would contain the `beforeStep` and `afterStep` specs which are arrays containing names of steps
+which the user plans on debugging. 
+
+_This functionality will only support named steps_
+
+```yaml
+  debug:
+    breakpoint:
+      onFailure: true
+      beforeStep: ["test-js"]
+      afterStep: ["push-to-registry"]
```

Comment:
Both `before` and `after` be specified, and they're both lists, correct?

Reply:
maybe renaming them to `beforeSteps` and `afterSteps` could make it clear that they are lists?

Reply:
done !